File: server/githubService.ts
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';

const DATA_DIR = path.resolve(__dirname, '../data');
if (!fs.existsSync(DATA_DIR)) {
  fs.mkdirSync(DATA_DIR, { recursive: true });
}

const CONNECTIONS_FILE = path.join(DATA_DIR, 'github_connections.json');
const TOKENS_FILE = path.join(DATA_DIR, 'github_tokens.json');

const GITHUB_CLIENT_ID = process.env.GITHUB_CLIENT_ID || '';
const GITHUB_CLIENT_SECRET = process.env.GITHUB_CLIENT_SECRET || '';
const GITHUB_CALLBACK_URL = process.env.GITHUB_CALLBACK_URL || 'http://localhost:8000/api/github/callback';
const FRONTEND_DEV_URL = process.env.FRONTEND_DEV_URL || 'http://localhost:5173/developer/github-connect';

interface StateEntry {
  role: 'main' | 'contributor';
  createdAt: number;
}

const oauthStates = new Map<string, StateEntry>();

function cleanupStates() {
  const now = Date.now();
  for (const [nonce, data] of oauthStates.entries()) {
    if (now - data.createdAt > 600000) {
      oauthStates.delete(nonce);
    }
  }
}

function loadConnections(): Record<string, any> {
  if (!fs.existsSync(CONNECTIONS_FILE)) {
    return {
      main: { connected: false, role: 'main' },
      contributor: { connected: false, role: 'contributor' },
    };
  }
  try {
    return JSON.parse(fs.readFileSync(CONNECTIONS_FILE, 'utf-8'));
  } catch {
    return {
      main: { connected: false, role: 'main' },
      contributor: { connected: false, role: 'contributor' },
    };
  }
}

function saveConnections(data: Record<string, any>) {
  fs.writeFileSync(CONNECTIONS_FILE, JSON.stringify(data, null, 2), 'utf-8');
}

function loadTokens(): Record<string, string> {
  if (!fs.existsSync(TOKENS_FILE)) return {};
  try {
    return JSON.parse(fs.readFileSync(TOKENS_FILE, 'utf-8'));
  } catch {
    return {};
  }
}

function saveTokens(tokens: Record<string, string>) {
  fs.writeFileSync(TOKENS_FILE, JSON.stringify(tokens, null, 2), 'utf-8');
}

export const GitHubService = {
  getStatus() {
    const connections = loadConnections();
    return {
      configured: Boolean(GITHUB_CLIENT_ID && GITHUB_CLIENT_SECRET),
      client_id_configured: Boolean(GITHUB_CLIENT_ID),
      callback_url: GITHUB_CALLBACK_URL,
      main: connections.main || { connected: false, role: 'main' },
      contributor: connections.contributor || { connected: false, role: 'contributor' },
      permissions_requested: [
        'Contents: Read & Write (code inspection, PRs, diff fixes)',
        'Pull Requests: Read & Write (collaborative contributor flow)',
        'User: Read (identify account & verify repo ownership)',
      ],
    };
  },

  createAuthUrl(role: 'main' | 'contributor') {
    if (!GITHUB_CLIENT_ID) {
      throw new Error('GITHUB_CLIENT_ID is not configured in server environment.');
    }
    cleanupStates();
    const stateNonce = crypto.randomBytes(24).toString('hex');
    oauthStates.set(stateNonce, { role, createdAt: Date.now() });

    const scope = 'repo read:user';
    const authUrl =
      `https://github.com/login/oauth/authorize` +
      `?client_id=${encodeURIComponent(GITHUB_CLIENT_ID)}` +
      `&redirect_uri=${encodeURIComponent(GITHUB_CALLBACK_URL)}` +
      `&scope=${encodeURIComponent(scope)}` +
      `&state=${encodeURIComponent(stateNonce)}` +
      `&allow_signup=false`;

    return { url: authUrl, state: stateNonce, role };
  },

  async handleCallback(code: string, state: string) {
    cleanupStates();
    const stateData = oauthStates.get(state);
    if (!stateData) {
      throw new Error('Invalid or expired OAuth state nonce. Please retry authorization.');
    }
    oauthStates.delete(state);
    const role = stateData.role;

    const tokenRes = await fetch('https://github.com/login/oauth/access_token', {
      method: 'POST',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        client_id: GITHUB_CLIENT_ID,
        client_secret: GITHUB_CLIENT_SECRET,
        code,
        redirect_uri: GITHUB_CALLBACK_URL,
      }),
    });

    const tokenData: any = await tokenRes.json();
    const accessToken = tokenData.access_token;
    if (!accessToken) {
      throw new Error(tokenData.error_description || 'Failed to obtain access token from GitHub.');
    }

    const userRes = await fetch('https://api.github.com/user', {
      headers: {
        'Authorization': `Bearer ${accessToken}`,
        'Accept': 'application/vnd.github+json',
      },
    });
    const userData: any = await userRes.json();
    const username = userData.login || '';

    const reposRes = await fetch('https://api.github.com/user/repos?sort=updated&per_page=30', {
      headers: {
        'Authorization': `Bearer ${accessToken}`,
        'Accept': 'application/vnd.github+json',
      },
    });
    const repos: any[] = (await reposRes.json()) || [];

    let selectedRepo = repos.find((r) =>
      ['code', 'codevault', 'codevault-pro', 'code_vault_pro'].includes((r.name || '').toLowerCase())
    );
    if (!selectedRepo && repos.length > 0) {
      selectedRepo = repos[0];
    }

    const repoFullName = selectedRepo?.full_name || `${username}/Code`;
    const repoId = selectedRepo?.id || 0;
    const defaultBranch = selectedRepo?.default_branch || 'main';
    const permissions = selectedRepo?.permissions || { push: true, pull: true, admin: false };

    // Save tokens strictly server-side
    const tokens = loadTokens();
    tokens[role] = accessToken;
    saveTokens(tokens);

    // Save connection metadata
    const connections = loadConnections();
    connections[role] = {
      connected: true,
      role,
      username,
      avatar_url: userData.avatar_url || '',
      full_name: repoFullName,
      repository_name: selectedRepo?.name || 'Code',
      repository_id: repoId,
      default_branch: defaultBranch,
      permissions: {
        push: Boolean(permissions.push),
        pull: Boolean(permissions.pull),
        admin: Boolean(permissions.admin),
      },
      connected_at: new Date().toISOString(),
      available_repos: repos.slice(0, 15).map((r) => ({
        id: r.id,
        full_name: r.full_name,
        name: r.name,
        private: r.private,
      })),
    };
    saveConnections(connections);

    return {
      role,
      username,
      repository: repoFullName,
      redirect_url: `${FRONTEND_DEV_URL}?status=connected&role=${role}&repo=${encodeURIComponent(repoFullName)}`,
    };
  },

  selectRepo(role: 'main' | 'contributor', repoFullName: string) {
    const connections = loadConnections();
    const target = connections[role];
    if (!target || !target.connected) {
      throw new Error(`No active connection found for ${role} repository.`);
    }
    target.full_name = repoFullName;
    target.repository_name = repoFullName.split('/').pop() || repoFullName;
    saveConnections(connections);
    return target;
  },

  disconnect(role: 'main' | 'contributor') {
    const tokens = loadTokens();
    delete tokens[role];
    saveTokens(tokens);

    const connections = loadConnections();
    connections[role] = { connected: false, role };
    saveConnections(connections);
    return { status: 'disconnected', role };
  },

  async testConnection(role: 'main' | 'contributor') {
    const tokens = loadTokens();
    const token = tokens[role];
    if (!token) return { valid: false, message: 'No server-side token configured.' };

    const connections = loadConnections();
    const repo = connections[role]?.full_name;
    if (!repo) return { valid: false, message: 'No repository selected.' };

    const res = await fetch(`https://api.github.com/repos/${repo}`, {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Accept': 'application/vnd.github+json',
      },
    });

    if (res.ok) {
      const data: any = await res.json();
      return {
        valid: true,
        repository: data.full_name,
        private: data.private,
        default_branch: data.default_branch,
        open_issues_count: data.open_issues_count,
        permissions: data.permissions,
      };
    }
    return {
      valid: false,
      status_code: res.status,
      message: 'GitHub API rejected token or repository not accessible.',
    };
  },
};

File: backend/services/github_service.py
import os
import json
import secrets
import time
import httpx
from typing import Dict, Any, Optional, List
from pathlib import Path
from backend.config import (
    GITHUB_CLIENT_ID,
    GITHUB_CLIENT_SECRET,
    GITHUB_CALLBACK_URL,
    FRONTEND_DEV_URL,
    DATA_DIR
)

CONNECTIONS_FILE = DATA_DIR / "github_connections.json"
TOKENS_FILE = DATA_DIR / "github_tokens.json"

# In-memory store for CSRF state nonces (nonce -> {role, created_at})
_OAUTH_STATES: Dict[str, Dict[str, Any]] = {}

def _cleanup_expired_states():
    now = time.time()
    expired = [s for s, data in _OAUTH_STATES.items() if now - data.get("created_at", 0) > 600]
    for s in expired:
        _OAUTH_STATES.pop(s, None)

def _load_connections() -> Dict[str, Any]:
    if not CONNECTIONS_FILE.exists():
        return {
            "main": {"connected": False, "role": "main"},
            "contributor": {"connected": False, "role": "contributor"}
        }
    try:
        with open(CONNECTIONS_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return {
            "main": {"connected": False, "role": "main"},
            "contributor": {"connected": False, "role": "contributor"}
        }

def _save_connections(data: Dict[str, Any]):
    with open(CONNECTIONS_FILE, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2)

def _load_tokens() -> Dict[str, str]:
    if not TOKENS_FILE.exists():
        return {}
    try:
        with open(TOKENS_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return {}

def _save_tokens(tokens: Dict[str, str]):
    with open(TOKENS_FILE, "w", encoding="utf-8") as f:
        json.dump(tokens, f, indent=2)

class GitHubService:
    @staticmethod
    def get_status() -> Dict[str, Any]:
        """Returns safe connection state for both repositories (No secret tokens exposed)."""
        connections = _load_connections()
        is_configured = bool(GITHUB_CLIENT_ID and GITHUB_CLIENT_SECRET)
        
        return {
            "configured": is_configured,
            "client_id_configured": bool(GITHUB_CLIENT_ID),
            "callback_url": GITHUB_CALLBACK_URL,
            "main": connections.get("main", {"connected": False, "role": "main"}),
            "contributor": connections.get("contributor", {"connected": False, "role": "contributor"}),
            "permissions_requested": [
                "Contents: Read & Write (code inspection, PRs, diff fixes)",
                "Pull Requests: Read & Write (collaborative contributor flow)",
                "User: Read (identify account & verify repo ownership)"
            ]
        }

    @staticmethod
    def create_auth_url(role: str) -> Dict[str, str]:
        """Generates secure official GitHub OAuth / App install URL with CSRF state."""
        if not GITHUB_CLIENT_ID:
            raise ValueError("GITHUB_CLIENT_ID is not configured in server environment.")
        
        _cleanup_expired_states()
        state_nonce = secrets.token_urlsafe(32)
        _OAUTH_STATES[state_nonce] = {
            "role": "contributor" if role == "contributor" else "main",
            "created_at": time.time()
        }

        # Request official GitHub authorization with minimum practical scope: repo + read:user
        # The user authenticates directly on official github.com
        scope = "repo read:user"
        auth_url = (
            f"https://github.com/login/oauth/authorize"
            f"?client_id={GITHUB_CLIENT_ID}"
            f"&redirect_uri={GITHUB_CALLBACK_URL}"
            f"&scope={scope}"
            f"&state={state_nonce}"
            f"&allow_signup=false"
        )
        return {"url": auth_url, "state": state_nonce, "role": role}

    @staticmethod
    async def handle_callback(code: str, state: str) -> Dict[str, Any]:
        """Exchanges authorization code for access token and fetches selected repo metadata."""
        _cleanup_expired_states()
        state_data = _OAUTH_STATES.pop(state, None)
        if not state_data:
            raise ValueError("Invalid or expired OAuth state nonce. Please retry authorization.")
        
        role = state_data.get("role", "main")

        # 1. Exchange code for access token via official GitHub OAuth endpoint
        async with httpx.AsyncClient(timeout=20.0) as client:
            token_resp = await client.post(
                "https://github.com/login/oauth/access_token",
                headers={"Accept": "application/json"},
                data={
                    "client_id": GITHUB_CLIENT_ID,
                    "client_secret": GITHUB_CLIENT_SECRET,
                    "code": code,
                    "redirect_uri": GITHUB_CALLBACK_URL,
                }
            )

            if token_resp.status_code != 200:
                raise ValueError("Failed to retrieve access token from GitHub.")

            token_json = token_resp.json()
            access_token = token_json.get("access_token")
            if not access_token:
                err_desc = token_json.get("error_description", "Unknown error from GitHub OAuth")
                raise ValueError(f"GitHub OAuth error: {err_desc}")

            # 2. Fetch authenticated user profile
            user_resp = await client.get(
                "https://api.github.com/user",
                headers={
                    "Authorization": f"Bearer {access_token}",
                    "Accept": "application/vnd.github+json"
                }
            )
            if user_resp.status_code != 200:
                raise ValueError("Failed to retrieve user profile from GitHub API.")
            
            user_data = user_resp.json()
            username = user_data.get("login", "")

            # 3. Fetch user repositories (to identify CodeVault repo)
            repos_resp = await client.get(
                "https://api.github.com/user/repos?sort=updated&per_page=30",
                headers={
                    "Authorization": f"Bearer {access_token}",
                    "Accept": "application/vnd.github+json"
                }
            )
            repos = repos_resp.json() if repos_resp.status_code == 200 else []

            # Match repository with name 'Code' or 'CodeVault' or pick top repo
            selected_repo = None
            for r in repos:
                r_name = r.get("name", "").lower()
                if r_name in ["code", "codevault", "codevault-pro", "code_vault_pro"]:
                    selected_repo = r
                    break
            
            if not selected_repo and repos:
                selected_repo = repos[0]

            repo_full_name = selected_repo.get("full_name") if selected_repo else f"{username}/Code"
            repo_id = selected_repo.get("id") if selected_repo else 0
            default_branch = selected_repo.get("default_branch", "main") if selected_repo else "main"
            permissions = selected_repo.get("permissions", {"push": True, "pull": True, "admin": False}) if selected_repo else {"push": True, "pull": True}

            # 4. Save tokens server-side only
            tokens = _load_tokens()
            tokens[role] = access_token
            _save_tokens(tokens)

            # 5. Save safe connection metadata
            connections = _load_connections()
            connections[role] = {
                "connected": True,
                "role": role,
                "username": username,
                "avatar_url": user_data.get("avatar_url", ""),
                "full_name": repo_full_name,
                "repository_name": selected_repo.get("name", "Code") if selected_repo else "Code",
                "repository_id": repo_id,
                "default_branch": default_branch,
                "permissions": {
                    "push": permissions.get("push", True),
                    "pull": permissions.get("pull", True),
                    "admin": permissions.get("admin", False)
                },
                "connected_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                "available_repos": [
                    {"id": r.get("id"), "full_name": r.get("full_name"), "name": r.get("name"), "private": r.get("private")}
                    for r in repos[:15]
                ]
            }
            _save_connections(connections)

            return {
                "role": role,
                "username": username,
                "repository": repo_full_name,
                "redirect_url": f"{FRONTEND_DEV_URL}?status=connected&role={role}&repo={repo_full_name}"
            }

    @staticmethod
    def select_repo(role: str, repo_full_name: str) -> Dict[str, Any]:
        """Allows selecting a specific repository from the authorized account."""
        connections = _load_connections()
        target = connections.get(role)
        if not target or not target.get("connected"):
            raise ValueError(f"No active connection found for {role} repository.")

        target["full_name"] = repo_full_name
        target["repository_name"] = repo_full_name.split("/")[-1] if "/" in repo_full_name else repo_full_name
        _save_connections(connections)
        return target

    @staticmethod
    def disconnect(role: str) -> Dict[str, Any]:
        """Revokes local server-side session and clears connection metadata."""
        tokens = _load_tokens()
        tokens.pop(role, None)
        _save_tokens(tokens)

        connections = _load_connections()
        connections[role] = {
            "connected": False,
            "role": role
        }
        _save_connections(connections)
        return {"status": "disconnected", "role": role}

    @staticmethod
    async def test_connection(role: str) -> Dict[str, Any]:
        """Verifies API access to the selected repository using server-side token."""
        tokens = _load_tokens()
        token = tokens.get(role)
        if not token:
            return {"valid": False, "message": "No server-side token configured."}

        connections = _load_connections()
        repo = connections.get(role, {}).get("full_name", "")
        if not repo:
            return {"valid": False, "message": "No repository selected."}

        async with httpx.AsyncClient(timeout=15.0) as client:
            resp = await client.get(
                f"https://api.github.com/repos/{repo}",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Accept": "application/vnd.github+json"
                }
            )
            if resp.status_code == 200:
                data = resp.json()
                return {
                    "valid": True,
                    "repository": data.get("full_name"),
                    "private": data.get("private"),
                    "default_branch": data.get("default_branch"),
                    "open_issues_count": data.get("open_issues_count"),
                    "permissions": data.get("permissions")
                }
            else:
                return {
                    "valid": False,
                    "status_code": resp.status_code,
                    "message": "GitHub API rejected token or repository not accessible."
                }
